force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,11 +185,61 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        
+        // Query to get the Account record matching the provided accountName
+        List<Account> accountList = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
+        
+        // Clone the list of opportunity names for further processing
+        List<String> oppNamesClone = oppNames.clone();
+
+        Account relatedAccount;
+    
+        // If the account does not exist, create a new Account record
+        if (accountList.isEmpty()) {
+            relatedAccount = new Account(Name = accountName);
+            insert relatedAccount;
+        } else {
+            // If the account exists, use the existing Account record
+            relatedAccount = accountList[0];
+        }
+
+        // Initialize a list to store the opportunities to be inserted/updated
+        List<Opportunity> opportunityList = new List<Opportunity>();
+
+        // Query to find existing opportunities that match the provided names
+        List<Opportunity> existingOppList = [SELECT Id, Name FROM Opportunity WHERE Name = :oppNames];
+
+        // If any existing opportunities are found, associate them with the related account
+        if (!(existingOppList.isEmpty())) {
+            for (Opportunity existingopp : existingOppList) {
+                existingopp.AccountId = relatedAccount.Id;
+                // Remove the name of the existing opportunity from the cloned list
+                oppNamesClone.remove(oppNamesClone.indexOf(existingopp.Name));
+                // Add the existing opportunity to the list to be upserted
+                opportunityList.add(existingopp);
+            }
+        }
+
+        // If there are still opportunity names in the cloned list, create new opportunities
+        if (!(oppNamesClone.isEmpty())) {
+            for (String opp : oppNamesClone) {
+                Opportunity newOpp = new Opportunity(
+                    Name = opp,
+                    AccountId = relatedAccount.Id,
+                    CloseDate = Date.today().addMonths(3),
+                    StageName = 'Prospecting'
+                );
+                // Add the new opportunity to the list to be upserted
+                opportunityList.add(newOpp);
+            }
+        }
+    
+        // Perform the upsert operation on the list of opportunities (inserts new and updates existing records)
+        upsert opportunityList;
```

Comment:
The way you solved this looks like it works fine. As you can guess, at this point in the class there is more than 1 way to solve this problem.